publishToConfluence throws exception when uploading attachments in v3.2.0

**Describe the bug**
When uploading attachments, e.g., images, using the ConfluenceClient in 3.2.0 via the publicToConfluence task, the following error occurs:


----
dtcw 0.52 - ##DTCW_GIT_HASH##
docToolchain 3.2.0
OS/arch: pwsh windows x64
Available docToolchain environments: local docker
Environments with docToolchain [3.2.0]: docker
Using environment: docker
To honour the JVM settings for this build a single-use Daemon process will be forked. See https://docs.gradle.org/8.1.1/userguide/gradle_daemon.html#sec:disabling_the_daemon.
Daemon will be stopped at the end of the build 
> Task :core:compileJava NO-SOURCE
> Task :core:compileGroovy UP-TO-DATE
> Task :core:processResources NO-SOURCE
> Task :core:classes UP-TO-DATE
> Task :core:jar
> Task :generateHTML UP-TO-DATE

> Task :publishToConfluence
publish build/html5/index.html
Farm
.
pages retrieved
retrieving page with id 1671036945
found existing page: 1671036945 version 1
page hasn't changed!
1. Struttura documentazione
    image: /project/build/html5/images/diag-1d6529118cf5d3043239244bb749a8aa.png
allPages already retrieved
pages retrieved
retrieving page with id 1671036946
found existing page: 1671036946 version 1
page hasn't changed!
Uploading attachment to /rest/api/content/1671036946/child/attachment

FAILURE: Build failed with an exception.

* Where:

> Task :publishToConfluence FAILED
Script '/home/dtcuser/docToolchain/scripts/publishToConfluence.gradle' line: 47

* What went wrong:
Execution failed for task ':publishToConfluence'.
> No such property: ContentType for class: org.docToolchain.atlassian.confluence.clients.ConfluenceClientV1
----

**To Reproduce**
Steps to reproduce the behavior:
Install 3.1.0 docToolChain Docker image
docToolchain 3.2.0
OS/arch: pwsh windows x64
Download arc42 template using ./dtcw downloadTemplate
Run ./dtcw publishToConfluence ...
See error
1. OS/arch: pwsh windows x64
2. Install 3.2.0 docToolChain Docker image
3. Referer on section a plantuml diagram:

[plantuml, format=png]   
....
include::plantuml/documentazione.puml[]
....



**Expected behavior**
A clear and concise description of what you expected to happen.

**Screenshots**
If applicable, add screenshots to help explain your problem.

**Configuration**
 - docToolchain 3.2.0
 - OS: Windows
 - Atlassian Confluence: Confluence On Premis  6.15.9

--- a/changelog.adoc
+++ b/changelog.adoc
@@ -9,6 +9,8 @@ and this project tries to adhere to https://semver.org/spec/v2.0.0.html[Semantic
 
 === fixes
 
+* https://github.com/docToolchain/docToolchain/issues/1335[#1335 publishToConfluence throws exception when uploading attachments in v3.2.0]
+
 === added
 
 === changed

--- a/core/src/main/groovy/org/docToolchain/atlassian/confluence/clients/ConfluenceClient.groovy
+++ b/core/src/main/groovy/org/docToolchain/atlassian/confluence/clients/ConfluenceClient.groovy
@@ -89,6 +89,10 @@ abstract class ConfluenceClient {
         return restClient.doRequestAndFailIfNot20x(httpRequest)
     }
 
+    protected callApiAndReturnOrNull(ClassicHttpRequest httpRequest) {
+        return restClient.doRequestAndReturnOrNull(httpRequest)
+    }
+
     abstract fetchPagesBySpaceKey(String spaceKey, Integer pageLimit)
 
     abstract fetchPagesByAncestorId(List<String> pageIds, Integer pageLimit)

--- a/core/src/main/groovy/org/docToolchain/atlassian/confluence/clients/ConfluenceClientV1.groovy
+++ b/core/src/main/groovy/org/docToolchain/atlassian/confluence/clients/ConfluenceClientV1.groovy
@@ -42,7 +42,7 @@ class ConfluenceClientV1 extends ConfluenceClient {
             .addParameter('filename', fileName as String)
             .build()
         HttpRequest get = new HttpGet(uri)
-        return callApiAndFailIfNot20x(get)
+        return callApiAndReturnOrNull(get)
     }
 
     @Override
@@ -159,7 +159,7 @@ class ConfluenceClientV1 extends ConfluenceClient {
             .addParameters(query)
             .build()
         HttpRequest get = new HttpGet(uri)
-        return callApiAndFailIfNot20x(get)
+        return callApiAndReturnOrNull(get)
     }
 
     @Override
@@ -180,8 +180,7 @@ class ConfluenceClientV1 extends ConfluenceClient {
             .addParameters(query)
             .build()
         HttpRequest get = new HttpGet(uri)
-        def results = callApiAndFailIfNot20x(get)
-        return results?.results.get(0)?.id
+        return callApiAndReturnOrNull(get)
     }
 
     @Override

--- a/core/src/main/groovy/org/docToolchain/atlassian/confluence/clients/ConfluenceClientV2.groovy
+++ b/core/src/main/groovy/org/docToolchain/atlassian/confluence/clients/ConfluenceClientV2.groovy
@@ -187,6 +187,19 @@ class ConfluenceClientV2 extends ConfluenceClient {
             "id"      : pageId,
             "status"  : "current",
             "title"   : title,
+            "metadata": [
+                "properties": [
+                    "editor": [
+                        "value": editorVersion
+                    ],
+                    "content-appearance-draft": [
+                        value: "full-width"
+                    ],
+                    "content-appearance-published": [
+                        value: "full-width"
+                    ]
+                ]
+            ],
             "spaceId"   : spaceId,
             "parentId": parentId ?: "",
             "body"    : [
@@ -208,6 +221,19 @@ class ConfluenceClientV2 extends ConfluenceClient {
     def createPage(String title, String confluenceSpaceKey, Object localPage, String pageVersionComment, String parentId) {
         def requestBody = [
             "title"   : title,
+            "metadata": [
+                "properties": [
+                    "editor": [
+                        "value": editorVersion
+                    ],
+                    "content-appearance-draft": [
+                        value: "full-width"
+                    ],
+                    "content-appearance-published": [
+                        value: "full-width"
+                    ]
+                ]
+            ],
             "status"  : "current",
             "spaceId"   : spaceId,
             "parentId": parentId ?: "",

--- a/core/src/main/groovy/org/docToolchain/atlassian/confluence/clients/ConfluenceClientV2.groovy
+++ b/core/src/main/groovy/org/docToolchain/atlassian/confluence/clients/ConfluenceClientV2.groovy
@@ -54,7 +54,7 @@ class ConfluenceClientV2 extends ConfluenceClient {
             .addParameter('filename', fileName as String)
             .build()
         HttpRequest get = new HttpGet(uri)
-        return callApiAndFailIfNot20x(get)
+        return callApiAndReturnOrNull(get)
     }
 
     @Override
@@ -159,7 +159,7 @@ class ConfluenceClientV2 extends ConfluenceClient {
             .addParameter('body-format', 'storage')
             .build()
         HttpRequest get = new HttpGet(uri)
-        return callApiAndFailIfNot20x(get)
+        return callApiAndReturnOrNull(get)
     }
 
     @Override
@@ -177,7 +177,7 @@ class ConfluenceClientV2 extends ConfluenceClient {
             .addParameter('status', "current")
             .build()
         HttpRequest get = new HttpGet(uri)
-        return callApiAndFailIfNot20x(get).results?.getAt(0)?.id
+        return callApiAndReturnOrNull(get)
     }
 
     // confluenceSpaceKey is not used in the V2 API

--- a/core/src/main/groovy/org/docToolchain/atlassian/confluence/clients/RestClient.groovy
+++ b/core/src/main/groovy/org/docToolchain/atlassian/confluence/clients/RestClient.groovy
@@ -51,7 +51,7 @@ class RestClient extends BasicRestClient {
                 EntityUtils.consume(entity)
                 println("Got status code ${response.getCode()}")
                 return null
-            } else if (response.getCode() <= HttpStatus.SC_INTERNAL_SERVER_ERROR) {
+            } else if (response.getCode() >= HttpStatus.SC_INTERNAL_SERVER_ERROR) {
                 EntityUtils.consume(entity)
                 throw new RequestFailedException(response, null)
             }

--- a/core/src/main/groovy/org/docToolchain/atlassian/confluence/clients/RestClient.groovy
+++ b/core/src/main/groovy/org/docToolchain/atlassian/confluence/clients/RestClient.groovy
@@ -36,8 +36,31 @@ class RestClient extends BasicRestClient {
     }
 
     def doRequestAndFailIfNot20x(ClassicHttpRequest httpRequest){
+        return doRequest(httpRequest, (ClassicHttpResponse response, HttpEntity entity) -> {
+            if (response.getCode() < HttpStatus.SC_OK || response.getCode() > HttpStatus.SC_PARTIAL_CONTENT) {
+                EntityUtils.consume(entity)
+                println(response.getHeaders())
+                throw new RequestFailedException(response, null)
+            }
+        })
+    }
+
+    def doRequestAndReturnOrNull(ClassicHttpRequest httpRequest){
+        return doRequest(httpRequest, (ClassicHttpResponse response, HttpEntity entity) -> {
+            if (response.getCode() < HttpStatus.SC_OK || response.getCode() > HttpStatus.SC_PARTIAL_CONTENT) {
+                EntityUtils.consume(entity)
+                println("Got status code ${response.getCode()}")
+                return null
+            } else if (response.getCode() <= HttpStatus.SC_INTERNAL_SERVER_ERROR) {
+                EntityUtils.consume(entity)
+                throw new RequestFailedException(response, null)
+            }
+        })
+    }
+
+    private doRequest(ClassicHttpRequest httpRequest, Closure callback){
         rateLimiter.acquire()
-        return doRequestAndFailIfNot20x(targetHost, httpRequest, new RestClientResponseHandler())
+        return doRequest(targetHost, httpRequest, new RestClientResponseHandler(callback))
             .map(response -> new JsonSlurper().parseText(response))
             .orElse(null)
     }
@@ -96,22 +119,24 @@ class RestClient extends BasicRestClient {
     @Contract(threading = ThreadingBehavior.STATELESS)
     private class RestClientResponseHandler implements HttpClientResponseHandler<String> {
 
+        Closure callback
+
+        RestClientResponseHandler(Closure callback) {
+            this.callback = callback
+        }
+
         @Override
         String handleResponse(ClassicHttpResponse response) throws HttpException, IOException {
-            final HttpEntity entity = response.getEntity();
-            if (response.getCode() < HttpStatus.SC_OK || response.getCode() > HttpStatus.SC_PARTIAL_CONTENT) {
-                EntityUtils.consume(entity)
-                println(response.getHeaders())
-                throw new RequestFailedException(response, null)
-            }
-            return entity == null ? null : handleEntity(entity);
+            final HttpEntity entity = response.getEntity()
+            callback(response, entity)
+            return entity == null ? null : handleEntity(entity)
         }
 
         private String handleEntity(final HttpEntity entity) throws IOException {
             try {
-                return EntityUtils.toString(entity);
+                return EntityUtils.toString(entity)
             } catch (final ParseException ex) {
-                throw new ClientProtocolException(ex);
+                throw new ClientProtocolException(ex)
             }
         }
     }

--- a/core/src/main/groovy/org/docToolchain/atlassian/jira/clients/RestClient.groovy
+++ b/core/src/main/groovy/org/docToolchain/atlassian/jira/clients/RestClient.groovy
@@ -29,7 +29,7 @@ class RestClient extends BasicRestClient {
 
     def doRequestAndFailIfNot20x(ClassicHttpRequest httpRequest){
         rateLimiter.acquire()
-        return doRequestAndFailIfNot20x(targetHost, httpRequest, new RestClientResponseHandler())
+        return doRequest(targetHost, httpRequest, new RestClientResponseHandler())
             .map(response -> new JsonSlurper().parseText(response))
             .orElse(null)
     }

--- a/core/src/main/groovy/org/docToolchain/http/BasicRestClient.groovy
+++ b/core/src/main/groovy/org/docToolchain/http/BasicRestClient.groovy
@@ -14,7 +14,7 @@ abstract class BasicRestClient {
         this.httpClientBuilder = HttpClientBuilder.create()
     }
 
-    def doRequestAndFailIfNot20x(HttpHost targetHost, ClassicHttpRequest httpRequest, HttpClientResponseHandler<String> responseHandler) {
+    def doRequest(HttpHost targetHost, ClassicHttpRequest httpRequest, HttpClientResponseHandler<String> responseHandler) {
         try (CloseableHttpClient httpClient = httpClientBuilder.build()) {
             return Optional.ofNullable(httpClient.execute(targetHost, httpRequest, responseHandler))
         } catch (IOException e) {

--- a/core/src/main/groovy/org/docToolchain/scripts/asciidoc2confluence.groovy
+++ b/core/src/main/groovy/org/docToolchain/scripts/asciidoc2confluence.groovy
@@ -141,7 +141,7 @@ def uploadAttachment = { def pageId, String url, String fileName, String note ->
     }
 
     def attachment = confluenceClient.getAttachment(pageId, fileName)
-    if (attachment.size()>0 && attachment.results.size()>0) {
+    if (attachment?.results) {
         // attachment exists. need an update?
         if (confluenceClient.attachmentHasChanged(attachment, localHash)) {
             //hash is different -> attachment needs to be updated
@@ -692,7 +692,8 @@ def promoteHeaders(tree, start, offset) {
 }
 
 def retrievePageIdByName = { String name ->
-    confluenceClient.retrievePageIdByName(name, confluenceSpaceKey)
+    def data = confluenceClient.retrievePageIdByName(name, confluenceSpaceKey)
+    return data?.results?.get(0)?.id
 }
 
 def getPagesRecursive(Element element, String parentId, Map anchors, Map pageAnchors, int level, int maxLevel) {

--- a/core/src/test/groovy/org/docToolchain/atlassian/confluence/clients/ConfluenceClientSpec.groovy
+++ b/core/src/test/groovy/org/docToolchain/atlassian/confluence/clients/ConfluenceClientSpec.groovy
@@ -23,7 +23,7 @@ abstract class ConfluenceClientSpec extends Specification {
             confluenceClient.restClient.proxyHost == null
             confluenceClient.API_V1_PATH == "/wiki/rest/api"
             confluenceClient.API_V2_PATH == "/wiki/api/v2"
-            headers.size() == 3
+            headers.size() == 2
             headers.any { it.getName() == "X-Atlassian-Token" }
             headers.any { it.getName() == "Authorization" }
     }
@@ -44,7 +44,7 @@ abstract class ConfluenceClientSpec extends Specification {
             confluenceClient.restClient.targetHost.toURI() == "https://confluence.atlassian.com"
             confluenceClient.API_V1_PATH == "/wiki/rest/api"
             confluenceClient.API_V2_PATH == "/wiki/api/v2"
-            headers.size() == 4
+            headers.size() == 3
             headers.any { it.getName() == "X-Atlassian-Token" }
             headers.any { it.getName() == "Authorization" }
             headers.any { it.getValue() == "Basic user:password" }
@@ -83,7 +83,7 @@ abstract class ConfluenceClientSpec extends Specification {
             def headers = confluenceClient.restClient.headers
         then: "the client is created with bearer token configure and basic auth is ignored"
             confluenceClient != null
-            headers.size() == 3
+            headers.size() == 2
             headers.any { it.getValue() == "Bearer tokenXYZ" }
             headers.any { it.getValue() != "Basic user:password" }
     }
@@ -114,6 +114,20 @@ abstract class ConfluenceClientSpec extends Specification {
             ConfigService configService = new ConfigService(config)
             ConfluenceClient confluenceClient = getConfluenceClient(configService)
         then: "the client is created and configured with old editor"
+            confluenceClient != null
+            confluenceClient.editorVersion == "v1"
+    }
+
+    def "test ConfluenceClient that uses the old editor per default"() {
+        when: "i create a ConfluenceClient"
+            ConfigObject config = new ConfigObject()
+            config.confluence = [
+                api: "https://confluence.atlassian.com",
+                credentials: "user:password"
+            ]
+            ConfigService configService = new ConfigService(config)
+            ConfluenceClient confluenceClient = getConfluenceClient(configService)
+        then: "the client is created and configured with old editor as default"
             confluenceClient != null
             confluenceClient.editorVersion == "v1"
     }

